### DISCUSSION TITLE
Missing tud_unmount_cb() and potential false tud_mount_cb() invoke

### DISCUSSION THREAD

**Discussion by Rocky04** [hathach/tinyusb#2201] [General]
@hathach 
I'm confused about the meaning of the `tud_umount_cb()` and `tud_mount_cb()` function. Without my suggested changes (#2171) the unmount can basically only be set for a self powered device because it's not triggered by a reset. The mounted state is set every time the configuration is set ([see here](https://github.com/hathach/tinyusb/blob/5cf94234d20cf74e1d6bb952ce1680b25f2568b3/src/device/usbd.c#L968)). ~~Right now that seems to be even the case when the host resets the configuration for some reason to be unconfigured (zero).~~ *Correction, so while the mounted callback is executed correctly there the callback for the unmounting event is missing.* Also the application needs to take care of the previous state. Like to know to resume to which power state when a suspend is abrogated. _If there is only one configuration `tud_mounted()` can be used to determine to which previous power state it should return. If there are more, the last configuration is needed too if the device is mounted. I'm not aware that this is mentioned somewhere._ 

To me that seems counterintuitive and that it's used in a kind of mixed way. If the mounted and unmounted state seems to be referring to the physical attachment of a self powered device to the USB bus then it would not make much sense to me to let TinyUSB handle the mounting state. For example the configuration descriptor tells the host if a device is self powered or not and how much current that can consume from USB. So if a self powered device wants to draw power so it maybe can recharge its batteries, the mounted state should be only called if a corresponding configuration descriptor is set. But TinyUSB has no knowledge of the exact configuration so that needs to be done in the custom code of the `tud_descriptor_configuration_cb()` function. _This of course depends on the view point._

~~**But even if that behavior is intended this way, `tud_mount_cb()` should only be invoked when the current configuration is non-zero. So that if it's zero `tud_umount_cb()` should be called instead.**~~ *Correction, this is already granted. Only the callback for the unmounting event is missing*.

_Edit:_
I think it should rather look like this:
![image](https://github.com/hathach/tinyusb/assets/5910569/f5418089-3811-436d-9c5e-928b6d0a2551)

_Edit 2:_
An example code is shown here #2202.

_Edit 3:_
The function [`tud_mounted()`](https://github.com/hathach/tinyusb/blob/5cf94234d20cf74e1d6bb952ce1680b25f2568b3/src/device/usbd.c#L351) to check if the device is mounted or not also indicate that the device should be handled as unmounted if the current configuration is zero and so that the callback should be invoked if the configuration is reset.

**Comment by Rocky04** [comment]
Currently all the examples seems to indicate that the device was previously mounted which can't be taken for granted. I think these should rather look like this:
```
void tud_resume_cb(void)
{
	blink_interval_ms = tud_mounted() ? BLINK_MOUNTED : BLINK_NOT_MOUNTED;
}
```

**Reply by hathach** [reply]
yeah you are right, mounted() and umounted() are actually for set_configured(). ~~I am kind of messed up with the naming~~. 

PS: Actually, I recall a bit about the naming:
- mounted = connected (attached) + configured. unmounted() 
- umounted: transition from mounted --> not operating mode. This should also include configure(index=0)

**Reply by Rocky04** [reply]
While I don't get what you mean with "mounted = connected (attached) + configured. unmounted()" the other for the umounted would be inline with my suggested change. So that this is invoked when the device went from a configured to an unconfigured state.

_Edit:_
I'm just a bit confused again with your wording, according to the code configured with an index of 0 would refer to the first configuration (cfg_num is 1). I assume you meant unconfigured (cfg_num is zero).
